### ISSUE TITLE
[Question] Is there a command to produce a preview image onSave?

### ISSUE THREAD

**Issue by onoufriosm** [TeselaGen/openVectorEditor#595]
@tnrich

Is there a way to produce an image preview during the onSave event (we want to store an image preview in our database on top of the sequenceData).

It could be the same `svg` (linear or circular) is computed to be displayed in the editor. Or any other format really like `png`.

Thanks!

Also, is this the right place to ask questions?

**Comment by tnrich** [comment]
Hey @onoufriosm yep this is the right place to ask! 

I unfortunately don't have the time to implement this feature itself but I think it is a great idea! If you're up for it, I can help you make a PR to get that functionality. 

Essentially it sounds like what you'll want to do is render either the linear or circular view and generate an image from it in the onSave handler. 

Here's where I imagine the custom code would hook in: 

```js
export const handleSave = props => (opts = {}) => {
  const {
    onSave,
    onSaveAs,
    readOnly,
    alwaysAllowSave,
    sequenceData,
    lastSavedIdUpdate
  } = props;
  const saveHandler = opts.isSaveAs ? onSaveAs || onSave : onSave;

  const updateLastSavedIdToCurrent = () => {
    lastSavedIdUpdate(sequenceData.stateTrackingId);
  };

//****call new code to generate the image here*****//

  const promiseOrVal =
    (!readOnly || alwaysAllowSave || opts.isSaveAs) &&
    saveHandler &&
    saveHandler(
      opts,
      tidyUpSequenceData(sequenceData, { annotationsAsObjects: true }),
      props,
      updateLastSavedIdToCurrent
    );

  if (promiseOrVal && promiseOrVal.then) {
    return promiseOrVal.then(updateLastSavedIdToCurrent);
  }
};
```
Here's the component you'll want to show (taken from the PrintDialog): 
```js
class ComponentToPrint extends React.Component {

  render() {
    const { editorName, circular } = this.props;

    return circular ? (
      <CircularView noInteractions editorName={editorName} />
    ) : (
      <LinearView noInteractions editorName={editorName} />
    );
  }
}

```

Once you've got the image showing up on the page, probably in a non-interactive modal that has a title like: "Saving Image...". Then you'll need to use some 3rd party lib like https://github.com/tsayen/dom-to-image to actually generate something you can save. 

Let me know if you have questions. I would maybe first see if you can just get a 3rd party tool to generate a png or svg of the circular or linear view. Once that works, make the dialog and have it save the image from the dialog. 

Good luck!

**Comment by onoufriosm** [comment]
@tnrich 

I've worked on this today and I've made progress. I can now produce a blob (and download it as png) of the circular/linear view. 

The one issue I have is that I can't use `ComponentToPrint` -> `Uncaught TypeError: Cannot read property 'apply' of undefined (redux)`. As a workaround for now I'm using the active view on the editor.

I can push my branch so that you see the work I've done. I was working off this repo and I don't have access to push. Do you want me to fork the repo and push or give me access to push on a separate branch on this repo?

Cheers.

**Comment by tnrich** [comment]
Hey @onoufriosm awesome! Ya you should fork the repo and make a pr from that branch. I can probably help you get past your error. Let me know when the pr is made :)

**Comment by onoufriosm** [comment]
https://github.com/TeselaGen/openVectorEditor/pull/598

**Comment by tnrich** [comment]
@onoufriosm were you able to figure it out from the changes I made?

**Comment by onoufriosm** [comment]
I'm going to work on this today. Thanks for the changes  though. I'll let you know if I have any issues

**Comment by tnrich** [comment]
Ok @onoufriosm this should be released in v10.1.63 and can be seen on the demo page here: http://teselagen.github.io/openVectorEditor/#/Editor?generatePng=true
